src/poprox_recommender/components/diversifiers/locality_calibration.py

```diff
@@ -25,13 +47,234 @@ def __call__(self, candidate_articles: ArticleSet, interest_profile: InterestPro
         article_indices = self.calibration(
             article_scores,
             candidate_articles.articles,
+            normalized_topic_prefs,
             normalized_locality_prefs,
-            self.theta,
+            self.theta_topic,
+            self.theta_local,
             topk=self.num_slots,
         )
-        return ArticleSet(articles=[candidate_articles.articles[int(idx)] for idx in article_indices])
 
-    def add_article_to_categories(self, rec_categories, article):
-        locality_list = extract_locality(article)
-        for locality in locality_list:
-            rec_categories[locality] = rec_categories.get(locality, 0) + 1
+        recommended_list = text_generation(candidate_articles.articles, article_indices, interest_profile)
+        return ArticleSet(articles=recommended_list)
+
+    def add_article_to_categories(self, rec_topics, article):
+        topics = extract_general_topics(article)
+        for topic in topics:
+            rec_topics[topic] = rec_topics.get(topic, 0) + 1
```

Comment:
Since the topics are `defaultdict`s, `rec_topics[topic]` will return 0 even `topic` isn't present in the dictionary yet, so here you can actually just do:
```suggestion
            rec_topics[topic] += 1
```
The same thing is true for localities in the method below